### PR TITLE
Very initial implementation of motion controls

### LINKED ISSUE #2635: Missing implementation of motion controls
While playing Super Mario odyssey the game asks you to shake your controller to do some moves, but when using a regular controller it is impossible.

<meta name="titleids"  content="0100000000010000,01009B500007C000,01007EF00011E000">

### PR DESCRIPTION
**Pull request by anirudhb**
Very initial implementation of motion controls. There's no calibration and currently the motion controls only support one controller.

This is my first time contributing to yuzu! 😃 

Closes #2635
GUI working!

Bugs:
 - [ ] Splatoon - only rotates horizontally, vertical doesn't work
 - [ ] ARMS - cannot move by tilting the controllers

What works:
 - Splatoon - horizontal rotation
 - SMO - basically perfect
 - ARMS - punching, curved punching, shielding using dual joycons (each joycon has its own motion source)

### CODE REVIEW

File: src/core/hle/service/hid/controllers/npad.h [hunk 1]
```diff
@@ -282,11 +301,14 @@ class Controller_NPad final : public ControllerBase {
...
         NPadGeneric pokeball_states;
-        NPadGeneric libnx; // TODO(ogniK): Find out what this actually is, libnx seems to only be
-                           // relying on this for the time being
-        INSERT_PADDING_BYTES(
-            0x708 *
-            6); // TODO(ogniK): SixAxis states, require more information before implementation
+        NPadGeneric libnx;   // TODO(ogniK): Find out what this actually is, libnx seems to only be
+                             // relying on this for the time being
+        SixAxisGeneric full; // used
+        SixAxisGeneric _handheld;   // stub
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/npad.h:307]: We don't use underscore prefixes on variable names.

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 2]
```diff
@@ -434,6 +456,12 @@ void Controller_NPad::OnUpdate(const Core::Timing::CoreTiming& core_timing, u8*
         libnx_entry.pad.l_stick = pad_state.l_stick;
         libnx_entry.pad.r_stick = pad_state.r_stick;
 
+        // Set this entry regardless of the controller type, for now.
+        if (motion_sensors[i] && sixaxis_sensor_enabled) {
+            auto& sensor = motion_sensors[i];
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/npad.cpp:461]:
```suggestion
            const auto& sensor = motion_sensors[i];
```

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 3]
```diff
@@ -338,6 +340,25 @@ void Controller_NPad::OnUpdate(const Core::Timing::CoreTiming& core_timing, u8*
             cur_entry.timestamp2 = cur_entry.timestamp;
         }
 
+        const std::array<SixAxisGeneric*, 6> controller_sixaxes{&npad.full,       &npad._handheld,
+                                                                &npad._left_dual, &npad._right_dual,
+                                                                &npad._left,      &npad._right};
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/npad.cpp:345]: Adding a comma at the end of the list, should allow clang-format to format this more nicely.

```suggestion
                                                                &npad._left,      &npad._right,};
```

File: src/core/hle/service/hid/controllers/npad.h [hunk 4]
```diff
@@ -236,6 +237,24 @@ class Controller_NPad final : public ControllerBase {
...
 
+    struct SixAxisState {
+        s64_le timestamp;
+        INSERT_PADDING_BYTES(8); // unknown
+        s64_le timestamp2;
+        Common::Vec3f accelerometer;
+        Common::Vec3f gyroscope;
+        INSERT_PADDING_BYTES(12); // unknown sensor data
+        std::array<Common::Vec3f, 3> orientation;
+        s64_le _always_one; // always 1
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/npad.h:248]: Ditto about underscore prefixes.

File: src/core/hle/service/hid/controllers/npad.h (hunk 1 above)
Comment (anirudhb) [reply src/core/hle/service/hid/controllers/npad.h:307]: Should I just remove this prefix then?

Comment (lioncash) [reply src/core/hle/service/hid/controllers/npad.h:307]: Yeah

**Comment by anirudhb** [comment]
I've addressed your nits.

**Comment by RealJohnGalt** [comment]
Working as expected on Linux with ds4drv + udp server.

**Comment by ogniK5377** [comment]
Has this been properly reversed?

File: src/core/hle/service/hid/controllers/npad.h [hunk 5]
```diff
@@ -236,6 +237,24 @@ class Controller_NPad final : public ControllerBase {
...
+        s64_le always_one; // always 1
+    };
+    static_assert(sizeof(SixAxisState) == 0x68, "SixAxisState is an invalid size");
+
+    struct SixAxisGeneric {
+        CommonHeader common;
+        std::array<SixAxisState, 17> sixaxis;
+    };
+    static_assert(sizeof(SixAxisGeneric) == 0x708, "SixAxisGeneric is an invalid size");
+
```
Comment (ogniK5377) [inline src/core/hle/service/hid/controllers/npad.h:257]:
```suggestion
    struct SixAxisState {
        s64_le timestamp{};
        INSERT_PADDING_BYTES(8); // unknown
        s64_le timestamp2{};
        Common::Vec3f accelerometer{};
        Common::Vec3f gyroscope{};
        INSERT_PADDING_BYTES(12); // unknown sensor data
        std::array<Common::Vec3f, 3> orientation{};
        s64_le always_one{1}; // always 1
    };
    static_assert(sizeof(SixAxisState) == 0x68, "SixAxisState is an invalid size");

    struct SixAxisGeneric {
        CommonHeader common;
        std::array<SixAxisState, 17> sixaxis;
    };
    static_assert(sizeof(SixAxisGeneric) == 0x708, "SixAxisGeneric is an invalid size");

```

File: src/input_common/udp/client.cpp [hunk 6]
```diff
@@ -52,7 +52,7 @@ class Socket {
     }
 
     void StartSend(const clock::time_point& from) {
-        timer.expires_at(from + std::chrono::seconds(3));
+        timer.expires_at(from);
```
Comment (jroweboy) [inline src/input_common/udp/client.cpp:55]: Why was this changed? IIRC the cemuhook compatible UDP programs are only expected to request updates once every 5 seconds, and the controller side will continually send updates for the next 5 seconds. It doesn't make sense to flood a request to the controller side.

Comment (anirudhb) [reply src/input_common/udp/client.cpp:55]: Noted, I'll revert it

Comment (RealJohnGalt) [reply src/input_common/udp/client.cpp]: FYI to observers: this was my fault from a noncompliant driver. Fixed driver side and all good with it reverted.

**Comment by anirudhb** [comment]
FYI addressed your nits @ogniK5377

**Comment by terencode** [comment]
Very nice, confirmed working on SMO with Linux [patched](https://patchwork.kernel.org/cover/11441643/) and [joycond](https://github.com/DanielOgorchock/joycond) + [joycond-cemuhook](https://github.com/joaorb64/joycond-cemuhook) on an unofficial Pro Controller.

I notice however multiple disconnection and connection logged by joycond-cemuhook while playing.
Is it a bug in joycond-cemuhook, an expected behaviour or something to fix here?

**Comment by anirudhb** [comment]
@terencode Same problem as with ds4drv, see TheDrHax/ds4drv-cemuhook#7

**Comment by terencode** [comment]
> @terencode Same problem as with ds4drv, see [TheDrHax/ds4drv-cemuhook#7](https://github.com/TheDrHax/ds4drv-cemuhook/pull/7)

Got it, will open PR/issue.

**Comment by callmenemo491** [comment]
clang failure... fix the code:
https://dev.azure.com/yuzu-emu/de968a8e-869e-4f5a-a0c9-0cb1b14a9357/_apis/build/builds/4502/logs/7

**Comment by anirudhb** [comment]
@callmenemo491 fixed! Thanks :)

**Comment by anirudhb** [comment]
Now ready for review, although have to refactor motion GUI into a separate dialog to go with @jroweboy 's coming changes to UI

**Comment by anirudhb** [comment]
Also, CI failures seem to be unrelated to this PR.

**Comment by ogniK5377** [comment]
Is there any reason for the submodules being out of date?

**Comment by anirudhb** [comment]
@fearlesstobi so would you rather this pr just be the hid stuff with config?

**Comment by anirudhb** [comment]
Closing in favor of a smaller, more targeted PR.

**Comment by FernandoS27** [comment]
Sorry for the inconvenience, it's just we need things setup in order.

**Comment by goldenx86** [comment]
It this dead?
